include/vrv/scoringupfunctor.h

```diff
@@ -0,0 +1,102 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        scoringupfunctor.h
+// Author:      Martha E. M. Thomae Elias
+// Created:     2024
+// Copyright (c) Authors and others. All rights reserved.
+/////////////////////////////////////////////////////////////////////////////
+///
+#ifndef __VRV_SCORINGUPFUNCTOR_H__
+#define __VRV_SCORINGUPFUNCTOR_H__
+
+#include "functor.h"
+
+namespace vrv {
+
+//----------------------------------------------------------------------------
+// ScoringUpFunctor
+//----------------------------------------------------------------------------
+
+/**
+ * This class...
+ */
+class ScoringUpFunctor : public Functor {
+public:
+    /**
+     * @name Constructors, destructors
+     */
+    ///@{
+    ScoringUpFunctor();
+    virtual ~ScoringUpFunctor() = default;
+    ///@}
+
+    /*
+     * Abstract base implementation
+     */
+    bool ImplementsEndInterface() const override { return true; }
+
+    /*
+     * Methods
+     */
+    /**
+     * @name: Divide the notes of a voice into sequences to be processed individualy
+     */
+    ///@{
+    std::vector<ArrayOfElementDurPairs> SubdivideSeq(const ArrayOfElementDurPairs &dursInVoiceSameMensur);
+    ///@}
+
+    /**
+     * @name: Find @dur.quality of notes (perfecta / imperfecta / altera)
+     */
+    ///@{
+    void FindDurQuals(const std::vector<ArrayOfElementDurPairs> &listOfSequences);
+    void FindDurQuals(const ArrayOfElementDurPairs &sequence);
+    ///@}
+
+    /**
+     * @name: Find the duration value of the note in minims
+     */
+    ///@{
+    double GetDurNumberValue(
+        std::pair<LayerElement *, data_DURATION> elementDurPair, bool followedByDot, LayerElement *nextElement);
+    ///@}
+
+    /**
+     * @name Apply principles of imperfection and alteration
+     */
+    ///@{
+    Note *ImperfectionAPP(const ArrayOfElementDurPairs &sequence);
+    Note *ImperfectionAPA(const ArrayOfElementDurPairs &sequence);
+    Note *Alteration(const ArrayOfElementDurPairs &sequence);
+    bool LeavePerfect(const ArrayOfElementDurPairs &sequence);
+    ///@}
+
+    /*
+     * Functor interface
+     */
+    ///@{
+    FunctorCode VisitLayer(Layer *layer) override;
+    FunctorCode VisitLayerElement(LayerElement *layerElement) override;
+    ///@}
+    ///
+
+protected:
+    //
+private:
+    //
+public:
+    std::map<std::string, int> m_mensuration;
```

Comment:
- If those variables are not used outside the functor, they should be made private too.
- Please define an enum capturing the different modes ("prolatio", "modusminor", ...) to avoid spelling them out as strings repeatedly which is error-prone. Then define `m_mensuration` as map from this enum to `int`.

Reply:
This is right, you should not use a string. Actually, wouldn't int members be simpler? Something like:
```
int m_modusMaior;
int m_modusMinor;
int m_tempus;
int m_prolatio;
```

Reply:
Thanks to both for the advice. Are the current changes good? If they are, we can mark this conversation as resolved too :)

Reply:
Yes, the changes are good, even thought it is a good practice to initialize the values in the constructor - I would agree that in a functor this might not be strictly necessary since the (re-)initialization might happen later down the processing tree.

Reply:
true, done!